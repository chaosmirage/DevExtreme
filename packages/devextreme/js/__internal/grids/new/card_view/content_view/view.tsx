/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable spellcheck/spell-checker */
import { compileGetter } from '@js/core/utils/data';
import { isDefined } from '@js/core/utils/type';
import { combined, computed, state } from '@ts/core/reactive/index';
import type { OptionsController } from '@ts/grids/new/card_view/options_controller';

import { ContentView as ContentViewBase } from '../../grid_core/content_view/view';
import type { DataObject } from '../../grid_core/data_controller/types';
import type { ContentViewProps } from './content_view';
import { ContentView as ContentViewComponent } from './content_view';

export class ContentView extends ContentViewBase<ContentViewProps> {
  // @ts-expect-error
  protected options: OptionsController;

  private readonly cardMinWidth = this.options.oneWay('cardMinWidth');

  private readonly rowHeight = state(0);

  private readonly cardsPerRow = computed(
    (width, cardMinWidth, pageSize, cardsPerRowProp) => {
      if (cardsPerRowProp !== 'auto') {
        return cardsPerRowProp;
      }

      function factors(n) {
        const res: number[] = [];
        for (let i = 1; i <= n; i += 1) if (n % i === 0) res.push(i);
        return res;
      }

      const result = factors(pageSize).reverse().find((cardsPerRow) => {
        const cardWidth = (width - 6 * (cardsPerRow - 1)) / cardsPerRow;
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return cardMinWidth! <= cardWidth;
      });

      return result ?? 1;
    },
    [this.width, this.cardMinWidth, this.dataController.pageSize, this.options.oneWay('cardsPerRow')],
  );

  private readonly virtualState = computed(
    (items, scrollTop, viewportHeight, rowHeight, cardsPerRow) => {
      const scrollHeight = (items.length / cardsPerRow) * rowHeight;

      const scrollBottom = scrollHeight - viewportHeight - scrollTop;

      const nonVisibleRowCountUp = Math.floor(scrollTop / rowHeight);
      const nonVisibleRowCountBottom = Math.floor(scrollBottom / rowHeight);

      const virtualTop = nonVisibleRowCountUp * rowHeight;
      const virtualBottom = nonVisibleRowCountBottom * rowHeight;

      const virtualItems = items.slice(
        nonVisibleRowCountUp * cardsPerRow,
        items.length - nonVisibleRowCountBottom * cardsPerRow,
      );

      return {
        virtualTop,
        virtualBottom,
        virtualItems,
      };
    },
    [this.items, this.scrollTop, this.viewportHeight, this.rowHeight, this.cardsPerRow],
  );

  protected override component = ContentViewComponent;

  protected override getProps() {
    return combined({
      ...this.getBaseProps(),
      contentProps: combined({
        items: this.items,
        // items: computed((virtualState) => virtualState.virtualItems, [this.virtualState]),
        fieldTemplate: this.options.template('fieldTemplate'),
        cardsPerRow: this.cardsPerRow,
        onRowHeightChange: this.rowHeight.update.bind(this.rowHeight),
        cardProps: combined({
          minWidth: this.cardMinWidth,
          maxWidth: this.options.oneWay('cardMaxWidth'),
          onClick: this.options.action('onCardClick'),
          onDblClick: this.options.action('onCardDblClick'),
          onHoverChanged: this.options.action('onCardHoverChanged'),
          onPrepared: this.options.action('onCardPrepared'),
          cover: combined({
            imageExpr: computed(
              (imageExpr) => this.processExpr(imageExpr),
              [this.options.oneWay('cardCover.imageExpr')],
            ),
            altExpr: computed(
              (altExpr) => this.processExpr(altExpr),
              [this.options.oneWay('cardCover.altExpr')],
            ),
          }),
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          toolbar: this.options.oneWay('cardHeader.items') as any,
        }),
      }),
      virtualScrollingProps: combined({
        heightUp: 0,
        heightDown: 0,
        // heightUp: computed((virtualState) => virtualState.virtualTop, [this.virtualState]),
        // heightDown: computed((virtualState) => virtualState.virtualBottom, [this.virtualState]),
      }),
    });
  }

  private processExpr<T>(
    expr: T | ((data: DataObject) => T) | undefined,
  ): ((data: DataObject) => T) | undefined {
    if (!isDefined(expr)) {
      return undefined;
    }
    // @ts-expect-error
    return compileGetter(expr);
  }
}
