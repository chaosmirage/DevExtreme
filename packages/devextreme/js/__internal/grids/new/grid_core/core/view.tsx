/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-this-alias */
/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable max-classes-per-file */
/* eslint-disable spellcheck/spell-checker */
import type { Subscription, SubsGets } from '@ts/core/reactive/index';
import { toSubscribable } from '@ts/core/reactive/index';
import { Component, type ComponentType, render } from 'inferno';
import { hydrate } from 'inferno-hydrate';
// import { renderToString } from 'inferno-server';

export abstract class View<T extends {}> {
  private inferno: undefined | ComponentType;

  private props?: T;

  private firstRender = true;

  protected abstract component: ComponentType<T>;

  protected abstract getProps(): SubsGets<T>;

  public render(root: Element): Subscription {
    const ViewComponent = this.component;
    return toSubscribable(this.getProps()).subscribe((props: T) => {
      this.props = props;
      if (this.firstRender) {
        this.firstRender = false;
        // @ts-expect-error
        hydrate(<ViewComponent {...props}/>, root);
      } else {
        // @ts-expect-error
        render(<ViewComponent {...props}/>, root);
      }
    });
  }

  public asInferno(): ComponentType {
    // eslint-disable-next-line no-return-assign
    return this.inferno ??= this._asInferno();
  }

  private _asInferno() {
    const view = this;

    interface State {
      props: T;
    }

    return class InfernoView extends Component<{}, State> {
      private subscription?: Subscription;

      private readonly viewProps: SubsGets<T>;

      constructor() {
        super();
        this.viewProps = view.getProps();
        this.state = {
          props: view.getProps().unreactive_get(),
        };
      }

      public render(): JSX.Element | undefined {
        const ViewComponent = view.component;
        // @ts-expect-error
        return <ViewComponent {...this.state!.props}/>;
      }

      public componentDidMount(): void {
        this.subscription = toSubscribable(this.viewProps).subscribe((props) => {
          this.state ??= {
            props,
          };

          if (this.state.props !== props) {
            this.setState({ props });
          }
        });
      }

      public componentWillUnmount(): void {
        this.subscription!.unsubscribe();
      }
    };
  }
}
