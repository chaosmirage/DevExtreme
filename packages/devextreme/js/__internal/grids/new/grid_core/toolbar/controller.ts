/* eslint-disable @typescript-eslint/no-dynamic-delete */
/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable spellcheck/spell-checker */
import type {
  MaybeSubscribable, Subscribable, Subscription, SubsGets,
} from '@ts/core/reactive/index';
import { computed, state, toSubscribable } from '@ts/core/reactive/index';

import { OptionsController } from '../options_controller/options_controller';
import { DEFAULT_TOOLBAR_ITEMS } from './defaults';
import type { PredefinedToolbarItem, ToolbarItem, ToolbarItems } from './types';
import { normalizeToolbarItems } from './utils';

export class ToolbarController {
  private readonly itemSubscriptions: Record<string, Subscription> = {};

  private readonly defaultItems = state<Record<string, PredefinedToolbarItem>>({});

  private readonly userItems: Subscribable<ToolbarItems | undefined>;

  public items: SubsGets<ToolbarItem[]>;

  public static dependencies = [OptionsController] as const;

  constructor(
    private readonly options: OptionsController,
  ) {
    this.userItems = this.options.oneWay('toolbar.items');
    this.items = computed(
      (defaultItems, userItems) => normalizeToolbarItems(
        Object.values(defaultItems),
        userItems,
        DEFAULT_TOOLBAR_ITEMS,
      ),
      [this.defaultItems, this.userItems],
    );
  }

  public addDefaultItem(
    item: MaybeSubscribable<PredefinedToolbarItem>,
  ): void {
    const itemObs = toSubscribable(item);
    // @ts-expect-error
    const { name } = itemObs.unreactive_get();

    this.itemSubscriptions[name] = itemObs.subscribe((item) => {
      this.defaultItems.updateFunc((oldDefaultItems) => ({
        ...oldDefaultItems,
        [item.name]: item,
      }));
    });
  }

  public removeDefaultItem(name: typeof DEFAULT_TOOLBAR_ITEMS[number]): void {
    this.defaultItems.updateFunc((oldDefaultItems) => {
      const defaultItems = { ...oldDefaultItems };
      delete defaultItems[name];
      return defaultItems;
    });
    this.itemSubscriptions[name].unsubscribe();
  }
}
