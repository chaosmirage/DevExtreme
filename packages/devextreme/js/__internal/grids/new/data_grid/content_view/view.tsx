/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable spellcheck/spell-checker */
import { combined } from '@ts/core/reactive/index';
import type { OptionsController } from '@ts/grids/new/card_view/options_controller';

import { ContentView as ContentViewBase } from '../../grid_core/content_view/view';
import type { ContentViewProps } from './content_view';
import { ContentView as ContentViewComponent } from './content_view';

export class ContentView extends ContentViewBase<ContentViewProps> {
  // @ts-expect-error
  protected options: OptionsController;

  protected override component = ContentViewComponent;

  protected override getProps() {
    return combined({
      ...this.getBaseProps(),
      contentProps: combined({
        items: this.items,
      }),
      virtualScrollingProps: combined({
        heightUp: 0,
        heightDown: 0,
      }),
    });
  }
}
