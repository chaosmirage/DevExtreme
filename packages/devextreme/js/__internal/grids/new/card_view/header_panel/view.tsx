/* eslint-disable spellcheck/spell-checker */
import type { SubsGets } from '@ts/core/reactive/index';
import { combined, computed } from '@ts/core/reactive/index';
import { ColumnsController } from '@ts/grids/new/grid_core/columns_controller/columns_controller';
import { View } from '@ts/grids/new/grid_core/core/view';

import type { Column } from '../../grid_core/columns_controller/types';
import type { HeaderPanelProps } from './header_panel';
import { HeaderPanel } from './header_panel';

export class HeaderPanelView extends View<HeaderPanelProps> {
  protected component = HeaderPanel;

  public static dependencies = [ColumnsController] as const;

  constructor(
    private readonly columnsController: ColumnsController,
  ) {
    super();
  }

  protected override getProps(): SubsGets<HeaderPanelProps> {
    return combined({
      columns: computed(
        (columns) => [...columns].sort((a, b) => a.visibleIndex - b.visibleIndex),
        [this.columnsController.columns],
      ),
      onMove: this.onMove.bind(this),
      onRemove: this.onRemove.bind(this),
      allowColumnReordering: this.columnsController.allowColumnReordering,
    });
  }

  public onRemove(column: Column): void {
    this.columnsController.columnOption(column, 'visible', !column.visible);
  }

  public onMove(column: Column, toIndex: number): void {
    // this.columnsController.columnOption(column, 'visible', true);
    this.columnsController.columnOption(column, 'visibleIndex', toIndex);
  }
}
