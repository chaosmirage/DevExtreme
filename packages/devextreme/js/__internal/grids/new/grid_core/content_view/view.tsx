/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable spellcheck/spell-checker */
import type dxScrollable from '@js/ui/scroll_view/ui.scrollable';
import type { ScrollEventInfo } from '@js/ui/scroll_view/ui.scrollable';
import { combined, computed, state } from '@ts/core/reactive/index';
import { ColumnsController } from '@ts/grids/new/grid_core/columns_controller/columns_controller';
import type { Column } from '@ts/grids/new/grid_core/columns_controller/types';
import { View } from '@ts/grids/new/grid_core/core/view';
import { DataController } from '@ts/grids/new/grid_core/data_controller/index';
import { ErrorController } from '@ts/grids/new/grid_core/error_controller/error_controller';
import { createRef } from 'inferno';

import { OptionsController } from '../options_controller/options_controller';

export abstract class ContentView<TProps extends {}> extends View<TProps> {
  private readonly isNoData = computed(
    (isLoading, items) => !isLoading && items.length === 0,
    [this.dataController.isLoading, this.dataController.items],
  );

  public readonly items = computed(
    (dataItems, columns: Column[]) => dataItems.map(
      (item) => this.columnsController.createDataRow(
        item,
        columns,
      ),
    ),
    [this.dataController.items, this.columnsController.visibleColumns],
  );

  public readonly scrollableRef = createRef<dxScrollable>();

  public loadingText = this.options.twoWay('loadPanel.message');

  protected readonly viewportHeight = state(0);

  protected readonly scrollTop = state(0);

  protected readonly width = state(0);

  public static dependencies = [
    DataController, OptionsController, ErrorController, ColumnsController,
  ] as const;

  constructor(
    protected readonly dataController: DataController,
    protected readonly options: OptionsController,
    protected readonly errorController: ErrorController,
    protected readonly columnsController: ColumnsController,

  ) {
    super();
  }

  protected getBaseProps() {
    return {
      loadPanelProps: computed(
        (visible, loadPanel) => ({
          ...loadPanel,
          visible,
        }),
        [
          this.dataController.isLoading,
          this.options.oneWay('loadPanel'),
        ],
      ),
      noDataTextProps: combined({
        text: this.options.oneWay('noDataText'),
        template: this.options.template('noDataTemplate'),
        visible: this.isNoData,
      }),
      errorRowProps: combined({
        enabled: this.options.oneWay('errorRowEnabled'),
        errors: this.errorController.errors,
      }),
      onWidthChange: this.width.update.bind(this.width),
      virtualScrollingProps: combined({
        heightUp: 0,
        heightDown: 0,
        // heightUp: computed((virtualState) => virtualState.virtualTop, [this.virtualState]),
        // heightDown: computed((virtualState) => virtualState.virtualBottom, [this.virtualState]),
      }),
      onViewportHeightChange: this.viewportHeight.update.bind(this.viewportHeight),
      scrollableRef: this.scrollableRef,
      scrollableProps: combined({
        onScroll: this.onScroll.bind(this),
        direction: 'both' as const,
        scrollTop: this.scrollTop,
        scrollByContent: this.options.oneWay('scrolling.scrollByContent'),
        scrollByThumb: this.options.oneWay('scrolling.scrollByThumb'),
        showScrollbar: this.options.oneWay('scrolling.showScrollbar'),
        useNative: computed(
          (useNative) => (useNative === 'auto' ? undefined : useNative),
          [this.options.oneWay('scrolling.useNative')],
        ),
      }),
    };
  }

  private onScroll(e: ScrollEventInfo<unknown>): void {
    this.scrollTop.update(e.scrollOffset.top);
  }
}
