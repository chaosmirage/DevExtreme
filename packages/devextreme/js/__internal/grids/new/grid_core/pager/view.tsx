/* eslint-disable spellcheck/spell-checker */
import type { SubsGets } from '@ts/core/reactive/index';
import { combined, computed } from '@ts/core/reactive/index';

import { View } from '../core/view';
import { DataController } from '../data_controller/index';
import { OptionsController } from '../options_controller/options_controller';
import type { PagerProps } from './pager';
import { PagerView as Pager } from './pager';
import { calculatePageSizes, isVisible } from './utils';

export class PagerView extends View<PagerProps> {
  protected override component = Pager;

  public static dependencies = [DataController, OptionsController] as const;

  private readonly pageSizesConfig = this.options.oneWay('pager.allowedPageSizes');

  private readonly allowedPageSizes = computed(
    (pageSizesConfig, pageSize) => calculatePageSizes(
      this.allowedPageSizes?.unreactive_get(),
      pageSizesConfig,
      pageSize,
    ),
    [this.pageSizesConfig, this.dataController.pageSize],
  );

  private readonly visibleConfig = this.options.oneWay('pager.visible');

  private readonly visible = computed(
    (visibleConfig, pageCount) => isVisible(visibleConfig, pageCount),
    [this.visibleConfig, this.dataController.pageCount],
  );

  constructor(
    private readonly dataController: DataController,
    private readonly options: OptionsController,
  ) {
    super();
  }

  protected override getProps(): SubsGets<PagerProps> {
    return combined({
      allowedPageSizes: this.allowedPageSizes,
      visible: this.visible,
      pageIndex: computed(
        // TODO: fix the '??'
        (pageIndex) => (pageIndex ?? 0) + 1,
        [this.dataController.pageIndex],
      ),
      pageIndexChanged: (value): void => this.dataController.pageIndex.update(value - 1),
      pageSize: this.dataController.pageSize,
      pageSizeChanged: this.dataController.pageSize.update,
      gridCompatibility: false,
      pageCount: this.dataController.pageCount,
      showPageSizeSelector: this.options.oneWay('pager.showPageSizeSelector'),
      _skipValidation: true,
    });
  }
}
