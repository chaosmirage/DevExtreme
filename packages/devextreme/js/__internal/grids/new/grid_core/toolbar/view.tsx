/* eslint-disable spellcheck/spell-checker */
import type { SubsGets } from '@ts/core/reactive/index';
import { combined, computed } from '@ts/core/reactive/index';

import { View } from '../core/view';
import { OptionsController } from '../options_controller/options_controller';
import { ToolbarController } from './controller';
import { ToolbarView as Toolbar } from './toolbar';
import type { ToolbarProps } from './types';
import { isVisible } from './utils';

export class ToolbarView extends View<ToolbarProps> {
  protected override component = Toolbar;

  private readonly visibleConfig = this.options.oneWay('toolbar.visible');

  private readonly visible = computed(
    (visibleConfig, items) => isVisible(visibleConfig, items),
    [this.visibleConfig, this.controller.items],
  );

  public static dependencies = [ToolbarController, OptionsController] as const;

  constructor(
    private readonly controller: ToolbarController,
    private readonly options: OptionsController,
  ) {
    super();
  }

  protected override getProps(): SubsGets<ToolbarProps> {
    return combined({
      visible: this.visible,
      items: this.controller.items,
      disabled: this.options.oneWay('toolbar.disabled'),
    });
  }
}
