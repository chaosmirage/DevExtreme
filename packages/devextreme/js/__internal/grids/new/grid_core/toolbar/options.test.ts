/* eslint-disable spellcheck/spell-checker */
/* eslint-disable @typescript-eslint/dot-notation */
import { describe, expect, it } from '@jest/globals';

import type { Options } from '../options';
import { OptionsControllerMock } from '../options_controller/options_controller.mock';
import { ToolbarController } from './controller';
import { ToolbarView } from './view';

const createToolbarView = (options?: Options): {
  rootElement: HTMLElement;
  optionsController: OptionsControllerMock;
} => {
  const rootElement = document.createElement('div');
  const optionsController = new OptionsControllerMock(options ?? {
    toolbar: {
      visible: true,
    },
  });

  const toolbarController = new ToolbarController(optionsController);
  const toolbar = new ToolbarView(toolbarController, optionsController);

  toolbar.render(rootElement);

  return {
    rootElement,
    optionsController,
  };
};

describe('Options', () => {
  describe('visilbe', () => {
    describe('when it is \'true\'', () => {
      it('Toolbar should be visible', () => {
        const { rootElement } = createToolbarView({
          toolbar: {
            visible: true,
          },
        });

        expect(rootElement).toMatchSnapshot();
      });
    });

    describe('when it is \'false\'', () => {
      it('Toolbar should be hidden', () => {
        const { rootElement } = createToolbarView({
          toolbar: {
            visible: false,
          },
        });

        expect(rootElement).toMatchSnapshot();
      });
    });

    describe('when changing it to \'false\' at runtime', () => {
      it('Toolbar should be hidden', () => {
        const { rootElement, optionsController } = createToolbarView({
          toolbar: {
            visible: true,
          },
        });

        optionsController.option('toolbar.visible', false);

        expect(rootElement).toMatchSnapshot();
      });
    });

    describe('when changing it to \'true\' at runtime', () => {
      it('Toolbar should be visible', () => {
        const { rootElement, optionsController } = createToolbarView({
          toolbar: {
            visible: false,
          },
        });

        optionsController.option('toolbar.visible', true);

        expect(rootElement).toMatchSnapshot();
      });
    });
  });

  describe('items', () => {
    describe('when these are not set', () => {
      it('Toolbar should be hidden', () => {
        const { rootElement } = createToolbarView({
          toolbar: {
            items: [],
          },
        });

        expect(rootElement).toMatchSnapshot();
      });
    });

    describe('when these are set', () => {
      it('Toolbar should be visible', () => {
        const { rootElement } = createToolbarView({
          toolbar: {
            items: [{
              location: 'before',
              widget: 'dxButton',
              options: {
                text: 'button1',
              },
            }, {
              location: 'after',
              widget: 'dxButton',
              options: {
                text: 'button2',
              },
            }],
          },
        });

        expect(rootElement).toMatchSnapshot();
      });
    });
  });

  describe('disabled', () => {
    describe('when it is \'true\'', () => {
      it('Toolbar should be disabled', () => {
        const { rootElement } = createToolbarView({
          toolbar: {
            visible: true,
            disabled: true,
          },
        });

        expect(rootElement).toMatchSnapshot();
      });
    });

    describe('when it is \'false\'', () => {
      it('Toolbar should not be disabled', () => {
        const { rootElement } = createToolbarView({
          toolbar: {
            visible: true,
            disabled: false,
          },
        });

        expect(rootElement).toMatchSnapshot();
      });
    });
  });
});
