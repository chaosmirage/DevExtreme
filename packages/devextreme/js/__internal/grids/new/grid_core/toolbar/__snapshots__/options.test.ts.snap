// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options disabled when it is 'false' Toolbar should not be disabled 1`] = `
<div>
  <div
    class="dx-toolbar dx-widget dx-visibility-change-handler dx-collection dx-toolbar-mini"
    role="toolbar"
  >
    <div
      class="dx-toolbar-items-container"
    >
      <div
        class="dx-toolbar-before"
        role="presentation"
      />
      <div
        class="dx-toolbar-center"
        role="presentation"
      />
      <div
        class="dx-toolbar-after"
        role="presentation"
      />
    </div>
  </div>
</div>
`;

exports[`Options disabled when it is 'true' Toolbar should be disabled 1`] = `
<div>
  <div
    aria-disabled="true"
    class="dx-toolbar dx-widget dx-state-disabled dx-visibility-change-handler dx-collection dx-toolbar-mini"
    role="toolbar"
  >
    <div
      class="dx-toolbar-items-container"
    >
      <div
        class="dx-toolbar-before"
        role="presentation"
      />
      <div
        class="dx-toolbar-center"
        role="presentation"
      />
      <div
        class="dx-toolbar-after"
        role="presentation"
      />
    </div>
  </div>
</div>
`;

exports[`Options items when these are not set Toolbar should be hidden 1`] = `
<div>
  
</div>
`;

exports[`Options items when these are set Toolbar should be visible 1`] = `
<div>
  <div
    class="dx-toolbar dx-widget dx-visibility-change-handler dx-collection"
    role="toolbar"
  >
    <div
      class="dx-toolbar-items-container"
    >
      <div
        class="dx-toolbar-before"
        role="presentation"
      >
        <div
          class="dx-item dx-toolbar-item dx-toolbar-button"
        >
          <div
            class="dx-item-content dx-toolbar-item-content"
          >
            <div
              aria-label="button1"
              class="dx-widget dx-button dx-button-mode-contained dx-button-normal dx-button-has-text"
              role="button"
              tabindex="0"
            >
              <div
                class="dx-button-content"
              >
                <span
                  class="dx-button-text"
                >
                  button1
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="dx-toolbar-center"
        role="presentation"
      />
      <div
        class="dx-toolbar-after"
        role="presentation"
      >
        <div
          class="dx-item dx-toolbar-item dx-toolbar-button"
        >
          <div
            class="dx-item-content dx-toolbar-item-content"
          >
            <div
              aria-label="button2"
              class="dx-widget dx-button dx-button-mode-contained dx-button-normal dx-button-has-text"
              role="button"
              tabindex="0"
            >
              <div
                class="dx-button-content"
              >
                <span
                  class="dx-button-text"
                >
                  button2
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Options visilbe when changing it to 'false' at runtime Toolbar should be hidden 1`] = `
<div>
  
</div>
`;

exports[`Options visilbe when changing it to 'true' at runtime Toolbar should be visible 1`] = `
<div>
  <div
    class="dx-toolbar dx-widget dx-visibility-change-handler dx-collection dx-toolbar-mini"
    role="toolbar"
  >
    <div
      class="dx-toolbar-items-container"
    >
      <div
        class="dx-toolbar-before"
        role="presentation"
      />
      <div
        class="dx-toolbar-center"
        role="presentation"
      />
      <div
        class="dx-toolbar-after"
        role="presentation"
      />
    </div>
  </div>
</div>
`;

exports[`Options visilbe when it is 'false' Toolbar should be hidden 1`] = `
<div>
  
</div>
`;

exports[`Options visilbe when it is 'true' Toolbar should be visible 1`] = `
<div>
  <div
    class="dx-toolbar dx-widget dx-visibility-change-handler dx-collection dx-toolbar-mini"
    role="toolbar"
  >
    <div
      class="dx-toolbar-items-container"
    >
      <div
        class="dx-toolbar-before"
        role="presentation"
      />
      <div
        class="dx-toolbar-center"
        role="presentation"
      />
      <div
        class="dx-toolbar-after"
        role="presentation"
      />
    </div>
  </div>
</div>
`;
